wikipedia_cleanup/ar_template_predictor.py

```diff
@@ -0,0 +1,104 @@
+from tqdm.auto import tqdm
+
+from wikipedia_cleanup.predictor import Predictor
+
+tqdm.pandas()
+
+
+class AssociationRulesTemplatePredictor(Predictor):
+    def __init__(
+        self,
+        min_support: float = 0.05,
+        min_confidence: float = 0.8,
+        min_template_support: float = 0.001,
+        transaction_freq: str = "W",
+    ) -> None:
+        super().__init__()
+        self.min_support: float = min_support
+        self.min_confidence: float = min_confidence
+        self.min_template_support: float = min_template_support
+        self.transaction_freq: str = transaction_freq
+
+    def fit(
+        self, train_data: pd.DataFrame, last_day: datetime, keys: List[str]
+    ) -> None:
+        self.template_mapping: Dict[str, FrozenSet[str]] = (
+            train_data.groupby("infobox_key")["template"].apply(frozenset).to_dict()
+        )
+        df = train_data[
+            ["infobox_key", "value_valid_from", "template", "property_name"]
+        ].copy()
+        df["value_valid_from"] = pd.to_datetime(df["value_valid_from"])
+        df = (
+            df.groupby(
+                [
+                    "infobox_key",
+                    "template",
+                    pd.Grouper(key="value_valid_from", freq=self.transaction_freq),
+                ]
+            )["property_name"]
+            .progress_apply(frozenset)
+            .groupby("template")
+            .progress_apply(tuple)
+        )
+        df = df[df.apply(len) >= df.apply(len).sum() * self.min_template_support]
+        rules: Dict[str, Dict[str, Set[str]]] = collections.defaultdict(
+            lambda: collections.defaultdict(set)
+        )
+        for template, tl in tqdm(df.iteritems(), total=len(df)):
+            _, mined_rules = apriori(
+                tl,
+                min_support=self.min_support,
+                min_confidence=self.min_confidence,
+                max_length=2,
+            )
+            for rule in mined_rules:
+                rules[template][rule.rhs[0]].add(rule.lhs[0])
+        self.rules: Dict[str, Dict[str, FrozenSet[str]]] = {
+            template: {key: frozenset(value) for key, value in template_rules.items()}
+            for template, template_rules in rules.items()
+        }
+
+    def predict_timeframe(
+        self,
+        data_key: np.ndarray,
+        additional_data: np.ndarray,
+        columns: List[str],
+        first_day_to_predict: date,
+        timeframe: int,
+    ) -> bool:
+        if not (bool(len(additional_data)) and bool(len(data_key))):
+            return False
+        template = data_key[-1, columns.index("template")]
+        prop_name = data_key[-1, columns.index("property_name")]
+        lhss = self.rules.get(template, {}).get(prop_name, frozenset())
+        if not lhss:
+            return False
+        relevant_rows = additional_data[
+            additional_data[:, columns.index("value_valid_from")]
+            >= first_day_to_predict
+        ]
```

Comment:
You could do binary search on the column to get the first index and then take everything right of it. That may speed up the predictions. We do that in `predict`, `_get_data_until`